Refactor dmm_suite for readability

dearmochi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
<!-- Include a small to medium description of what your PR changes. Document every changes or this may delay review or even discourage maintainers from merging your PR! -->
Cleans up arguments, comments and paths of `/dmm_suite` (now `/datum/dmm_suite`) to increase readability should other developers want to take a look at it. Any suggestions to make it even more readable are appreciated!

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AffectedArc07 commented on code/modules/awaymissions/maploader/writer.dm:
```diff
@@ -5,28 +5,27 @@
-	//Check for illegal characters in file name... in a cheap way.
-	if(!((ckeyEx(map_name)==map_name) && ckeyEx(map_name)))
+/datum/dmm_suite
+	var/static/list/letter_digits = list(
```
Shouldnt this be in `dmm_suite.dm`

dearmochi replied:
You're right, in my mind I thought it'd make more sense to put this in `writer.dmm` because it's exclusively used there but I think for consistency all static stuff should be in the main file along with `var/static/quote` and the regexes

AffectedArc07 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -1,13 +1,13 @@
+/datum/dmm_suite
 	// These regexes are global - meaning that starting the maploader again mid-load will
 	// reset progress - which means we need to track our index per-map, or we'll
 	// eternally recurse
```
Same with these regexes

dearmochi replied:
Moved

AffectedArc07 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -470,12 +465,12 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
 // a new area type for each new ruin
-/dmm_suite/loaded_map
+/datum/dmm_suite/loaded_map
 	parent_type = /datum
```
This isnt needed anymore

dearmochi replied:
You mean the `parent_type` right?

AffectedArc07 replied:
Yeah. You don’t need to set `parent_type` to datum when it’s already a datum.

dearmochi replied:
Alright, removed

AffectedArc07 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +255,182 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
 // This ain't re-entrant, but we had this before the maploader update
-/dmm_suite/preloader
+/datum/dmm_suite/preloader
 	parent_type = /datum
```
Nor is this

dearmochi replied:
Removed.

Fox-McCloud commented:
I'm normally always for refactors--but this is a case of something so ancient and so untouched I'm afraid it'll break in ways we don't understand unless we're insanely careful.

We know it works, at the moment---dunno if we really want to touch it, aside from implementing TG's version of it.

@Crazylemon64 @AffectedArc07 your input is the most valuable here.

AffectedArc07 commented:
TG **mostly** killed off dmm_suite to turn it into `/datum/parsed_map`, which creates a parsed map on `/New()` of the datum, which I think is quite honestly awful from a design perspective. 

I would also like to point out that I suggested this as a joke in the discord just saying `hey someone refactor /dmm_suite` for me, so power to @dearmochi for actually doing this. I will say the same thing as I said in my initial review of this PR. Please make sure this is **thoroughly** tested, and maybe even request a TM. 

I think this is a good idea, but we **MUST** make sure its fully working before merge.

dearmochi commented:
https://github.com/ParadiseSS13/Paradise/blob/a640335c2986dc164dffd73f41f831e6e1a78c9c/code/modules/awaymissions/maploader/writer.dm#L106

Oof. During one of my mass replacements, I added a space in the coordinates to be written in `write_map`. That caused the regex to not recognize it and thus failing to load the map because it'd think it didn't have valid bounds. That is now fixed and after extensive testing it works again, as well as map loading (both away missions and smaller maps)

dearmochi commented:
I'm somewhat confident this now works at it should, but I'd recommend a TM with admins messing around in build mode to test out the saving/loading

farie82 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +243,181 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
-		//if others /turf are presents, simulates the underlays piling effect
+		// if others /turf are presents, simulates the underlays piling effect
 		index = first_turf_index + 1
 		while(index <= members.len - 1) // Last item is an /area
```
Best to cache the length prior to the while loop

farie82 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +243,181 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
-//Instance an atom at (x,y,z) and gives it the variables in attributes
-/dmm_suite/proc/instance_atom(path,list/attributes, x, y, z)
+// Instance an atom at (x, y, z) and gives it the variables in attributes
+/datum/dmm_suite/proc/instance_atom(path, list/attributes, x = 0, y = 0, z = 0)
```
This should never be called with default values so best to not add them either

farie82 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +243,181 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
 
 		else
-			instance = new path (T)//first preloader pass
+			instance = new path (T) // first preloader pass
```
no space after `path` here

farie82 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +243,181 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
-/dmm_suite/proc/trim_text(what as text,trim_quotes=0)
+// text trimming (both directions) helper proc
+// optionally removes quotes before and after the text (for variable name)
+/datum/dmm_suite/proc/trim_text(what = "", trim_quotes = FALSE)
```
The `what` should not be defaulted either. Calling this empty doesn't make much sense

farie82 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +243,181 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
-/dmm_suite/proc/find_next_delimiter_position(text as text,initial_position as num, delimiter=",",opening_escape=quote,closing_escape=quote)
+// find the position of the next delimiter, skipping whatever is comprised between opening_escape and closing_escape
+// returns 0 if reached the last delimiter
+/datum/dmm_suite/proc/find_next_delimiter_position(text = "", initial_position = 0, delimiter = ",", opening_escape = quote, closing_escape = quote)
```
`text` here should not have a default

farie82 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -255,186 +243,181 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new())
-
+// build a list from variables in text form (e.g {var1="derp"; var2; var3=7} => list(var1="derp", var2, var3=7))
+// return the filled list
+/datum/dmm_suite/proc/readlist(text = "", delimiter = ",")
```
Same here. text shouldn't have a default

farie82 commented on code/modules/awaymissions/maploader/writer.dm:
```diff
@@ -54,96 +38,110 @@
 		CRASH("No templates found!")
-	var/key_length = round/*floor*/(log(letter_digits.len,templates.len-1)+1)
+
+	var/key_length = round/*floor*/(log(letter_digits.len, templates.len - 1) + 1)
```
Best to put this comment here behind the actual code. Or above it. Placing it in between... jeesh how odd

farie82 commented on code/modules/awaymissions/maploader/writer.dm:
```diff
@@ -54,96 +38,110 @@
-	if(flags & DMM_USE_JSON)
-		use_json = 1
+/datum/dmm_suite/proc/make_template(turf/model, flags = 0)
+	var/use_json = flags & DMM_USE_JSON
```
`use_json` will now have the value of `DMM_USE_JSON` or 0. It will work but it's best to make this a boolean value just for consistency

farie82 commented on code/modules/awaymissions/maploader/writer.dm:
```diff
@@ -155,65 +153,68 @@
-		if(yeah.len)
-			var/json_stuff = json_encode(yeah)
+		to_encode -= "type"
+		if(to_encode.len)
```
`length` is preferred over `.len`. It's both faster and null pointer safe

farie82 commented on code/modules/awaymissions/maploader/writer.dm:
```diff
@@ -155,65 +153,68 @@
 
-
-/dmm_suite/proc/get_model_key(var/which as num, var/key_length as num)
+/datum/dmm_suite/proc/get_model_key(which = 0, key_length = 0)
```
These should not have default values

Fox-McCloud commented:
If this breaks something, @AffectedArc07 has a token to come after you @dearmochi 

😛
